### PR TITLE
[Train] Immediately fail if application errors on any worker

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: Amog Kamsetty <amogkamsetty@yahoo.com>
Resolves https://discuss.ray.io/t/torch-trainer-gets-stuck/7447

Previously with Ray Train, if 1 worker raises an Exception in the application code, this Exception is not raised in the driver until all workers have finished executing the training function. This can lead to hanging behaviors where other workers are waiting on a collective/`torch.distributed.barrier()` call for example.

With this PR, if any worker fails in the application code, the entire training job is immediately terminated and the exception is raised on the driver.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/_internal/session.py [hunk 1]
```diff
@@ -235,6 +261,13 @@ def _auto_fill_checkpoint_metrics(self, result: dict) -> dict:
         result.update(auto_filled_metrics)
         return result
 
+    def _report_thread_runner_error(self, block=False):
+        try:
+            e = self.error_queue.get(block=block, timeout=ERROR_FETCH_TIMEOUT)
+            raise StartTraceback from skip_exceptions(e)
```
Comment (krfricke) [inline python/ray/train/_internal/session.py:267]: See https://github.com/ray-project/ray/pull/28333

```suggestion
            raise StartTraceback from e
```

File: python/ray/train/constants.py [hunk 2]
```diff
@@ -35,6 +35,10 @@
 # new results after signaling the training function to continue.
 RESULT_FETCH_TIMEOUT = 0.2
 
+# Time between Session.get_next checks for fetching exceptions raised by the training
+# function.
+ERROR_FETCH_TIMEOUT = 1
```
Comment (krfricke) [inline python/ray/train/constants.py:40]: Should we also move this to `ray.air.constants` and use the same in function trainable and ray train?

File: python/ray/train/tests/test_trainer.py [hunk 3]
```diff
@@ -472,8 +472,8 @@ def train_func_checkpoint():
 @pytest.mark.parametrize(
     "logdir",
     [
-        None,
-        "/tmp/test/trainer/test_persisted_checkpoint",
+        # None,
+        # "/tmp/test/trainer/test_persisted_checkpoint",
```
Comment (krfricke) [inline python/ray/train/tests/test_trainer.py:476]: Revert comment?

**Review by krfricke — COMMENTED** [review]
LGTM, minor comments

File: python/ray/train/_internal/session.py (hunk 1 above)
Comment (amogkam) [reply python/ray/train/_internal/session.py]: Updated!

File: python/ray/train/constants.py (hunk 2 above)
Comment (amogkam) [reply python/ray/train/constants.py]: Good point, updated!

File: python/ray/train/tests/test_trainer.py (hunk 3 above)
Comment (amogkam) [reply python/ray/train/tests/test_trainer.py]: Ah good catch, reverted!

**Comment by amogkam** [comment]
@krfricke can you review again please?

**Review by krfricke — APPROVED** [review]
LGTM

**Pull request merged**